### PR TITLE
enhance transpose simplifier pass

### PR DESCRIPTION
**Pull request by wyzero**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc [hunk 1]
```diff
@@ -882,6 +922,141 @@ LogicalResult reverseBinaryOpsIfBeneficial(Block* block, bool& changed) {
...
+      for (auto& transposedUser : transposedUsers) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "\t\t transposedUser: " << *transposedUser << "\n");
+      }
+      for (auto user : curOp->getResult(0).getUsers()) {
+        LLVM_DEBUG(llvm::dbgs() << "\tuser of curOp: " << *user << "\n");
+        if (isa<tensor::DimOp, shape::ShapeOfOp, mhlo::TransposeOp>(user))
+          continue;
+        if (transposedUsers.find(user) != transposedUsers.end()) continue;
+        needReverse = true;
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc:989]: nit: break?

File: tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc [hunk 2]
```diff
@@ -882,6 +922,141 @@ LogicalResult reverseBinaryOpsIfBeneficial(Block* block, bool& changed) {
...
+        // transpose can be folded for const op.
+        if (dyn_cast_or_null<mhlo::ConstantOp>(operandDefiningOp)) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "\t\tskip propagate for operand #" << en.index()
+                     << " due to operand op is const op\n");
+          continue;
+        }
+        bool needInsertTransposeOp =
+            tryUpdate(operandDefiningOp, curOp, operandPerm);
+        if (needInsertTransposeOp) {
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc:1034]: so the current 'cost' is number of transpose op? Not the data amount transposed ?

File: tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc (hunk 1 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc]: It's the last statement of the for loop body, thus I think it's not necessary to add a break here.

File: tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc (hunk 2 above)
Comment (qiuxiafei) [reply tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc]: Anyway, let's start here. Refine later, maybe.

**Review by qiuxiafei — APPROVED** [review]

**Pull request merged**